index/types.go

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+// Tag represents a metric tag.
+type Tag struct {
+	Key   []byte
+	Value []byte
+}
+
+// ID represents a metric identifier.
+type ID []byte
+
+// Index represents a reverse index.
+type Index interface {
+	// Insert inserts the given ID with provided tags.
+	Insert(ID, []Tag) error
```

Comment:
Do we want to provide a way to associate data with an `ID`? For instance, for `m3db` we may eventually want to hold metadata about the series in the index. 

Some thoughts I had here are (names need some work):

1. Associate a byte slice with the ID and provide a method to return that data.
```
Insert(ID, []Tag, []byte) error
...
FetchPayload(FetchOptions) []{ID, Payload}
```

2. Take a more generic, document oriented approach.

```
type Document struct {
    ID ID
    Tags []Tag // these are indexed
    Fields []Tag // these are not indexed but are returned
}

Insert(Document) error

Fetch(FetchOptions) []Document
```

Reply:
Sure thing. I'll make `Document` an interface with the primitives we need.

Reply:
I wonder if it even needs to be an interface, I think a struct might be sufficient.

Reply:
The only reason for the interface is to cache additional state (e.g. we'd want to avoid hashing the bytes multiple times)